### PR TITLE
feat(PSDK-670): Support external wallet imports, wallet imports from CDP Python SDK

### PR DESCRIPTION
**Pull request by derek-cb**
### What changed? Why?

- Support external 1-of-1 wallet import via BIP-39 mnemonic phrases (12, 15, 18, 21, or 24 words)
- Support import of wallet data exported from CDP Python SDK
- Renamed saveSeed() to saveSeedtoFile(), and deprecated the former
- Renamed loadSeed() to loadSeedFromFile(), and deprecated the former
- Added retry mechanism to address.listTransactions() E2E test for better resiliency

### Testing

Tested via:
- Import of 12, 15, 18, 21, and 24-word mnemonic seed phrases
- Export of wallets generated via mnemonic seed phrases
- Import of wallets generated via mnemonic seed phrases, that were then exported 
- Successfully imported wallet exported from CDP Python SDK `v0.12.0`
- Successfully imported wallet exported from CDP NodeJS SDK `v0.11.0`

### CODE REVIEW

File: src/coinbase/wallet.ts [hunk 1]
```diff
@@ -130,31 +136,82 @@ export class Wallet {
...
    * @returns The imported Wallet.
-   * @throws {ArgumentError} If the Wallet ID is not provided.
-   * @throws {ArgumentError} If the seed is not provided.
+   * @throws {ArgumentError} If the seed phrase is not provided or invalid.
+   * @throws {ArgumentError} If the seed phrase is not 12, 15, 18, 21, or 24 words.
    * @throws {APIError} If the request fails.
    */
-  public static async import(data: WalletData): Promise<Wallet> {
-    if (!data.walletId) {
+  public static async importFromMnemonicSeedPhrase(mnemonicSeedPhrase: string): Promise<Wallet> {
```
Comment (John-peterson-coinbase) [inline src/coinbase/wallet.ts:147]: `importFromMnemonicSeedPhrase` is quite verbose. The PRD specifies `import` method should take mnemonic seed as input. 

Did we discuss with product the deprecation and method name changes?

Comment (derek-cb) [reply src/coinbase/wallet.ts]: Resolved via offline convo. Changes appear in latest commit

File: src/coinbase/wallet.ts [hunk 2]
```diff
@@ -130,31 +136,67 @@ export class Wallet {
...
+   * @param data - The data used to import the wallet:
+   *   - If WalletData: Must contain walletId (or wallet_id) and seed.
+   *     Allows for the loading of an existing CDP wallet into CDP.
+   *   - If MnemonicSeedPhrase: Must contain a valid BIP-39 mnemonic phrase (12, 15, 18, 21, or 24 words).
+   *     Allows for the import of an external wallet into CDP as a 1-of-1 wallet.
+   * @returns A Promise that resolves to the loaded Wallet instance
+   * @throws {ArgumentError} If the data format is invalid.
    * @throws {ArgumentError} If the seed is not provided.
-   * @throws {APIError} If the request fails.
+   * @throws {ArgumentError} If the mnemonic seed phrase is invalid.
```
Comment (jazz-cb) [inline src/coinbase/wallet.ts:149]: shouldnt we still keep the `throws` comment?

File: src/coinbase/types.ts [hunk 3]
```diff
@@ -812,14 +812,87 @@ export enum FundOperationStatus {
...
+ * property names for compatibility with older versions of the Python SDK.
  */
-export type WalletData = {
-  walletId: string;
+export interface WalletData {
+  /**
+   * The CDP wallet ID in either camelCase or snake_case format, but not both.
+   */
+  walletId?: string;
+  wallet_id?: string;
```
Comment (0xRAG) [inline src/coinbase/types.ts:823]: One option to try to enforce mutual exclusivity is to use `never`:
```ts
type WalletData = {
    walletId: string;
    wallet_id?: never;
} | {
    walletId?: never;
    wallet_id: string;
}

const walletData: WalletData = {
    walletId: '123',
    wallet_id: '123'
}
```

Now TypeScript will error with a message:
```
Type '{ walletId: string; wallet_id: string; }' is not assignable to type 'WalletData'.
  Types of property 'wallet_id' are incompatible.
    Type 'string' is not assignable to type 'undefined'
```

Admittedly, the error message could be clearer, but it's difficult to customize the message without reaching for a type validation library like Zod, which might be overkill here.

File: src/coinbase/types.ts [hunk 4]
```diff
@@ -812,14 +812,87 @@ export enum FundOperationStatus {
...
 
 /**
- * The Wallet Data type definition.
- * The data required to recreate a Wallet.
+ * Interface representing wallet data, with support for both camelCase and snake_case
+ * property names for compatibility with older versions of the Python SDK.
  */
-export type WalletData = {
-  walletId: string;
+export interface WalletData {
```
Comment (0xRAG) [inline src/coinbase/types.ts:818]: Curious on reasoning to convert this from a `type` to an `interface`?

File: src/coinbase/wallet.ts [hunk 5]
```diff
@@ -130,31 +136,67 @@ export class Wallet {
...
+        throw new ArgumentError("BIP-39 mnemonic seed phrase must be provided");
+      }
+
+      if (!validateMnemonic(data.mnemonicPhrase, wordlist)) {
+        throw new ArgumentError("Invalid BIP-39 mnemonic seed phrase");
+      }
+
+      // Convert mnemonic phrase to seed
+      const seedBuffer = mnemonicToSeedSync(data.mnemonicPhrase);
+      const seed = hexlify(seedBuffer).slice(2); // remove 0x prefix
```
Comment (0xRAG) [inline src/coinbase/wallet.ts:166]: Probably out-of-scope for this PR, but I would love to switch to `viem` and/or [Ox](https://oxlib.sh/).

That said, we do already have `viem` installed as a dependency, so we could use [`toHex`](https://viem.sh/docs/utilities/toHex#tohex) instead!

File: src/coinbase/wallet.ts [hunk 6]
```diff
@@ -130,31 +136,67 @@ export class Wallet {
...
+
+      // Create wallet using the provided seed
+      const wallet = await Wallet.createWithSeed({
+        seed: seed,
```
Comment (0xRAG) [inline src/coinbase/wallet.ts:170]: Nit: Shorthand syntax can be used here (this should be enforced ideally by the linter / formatter, I'll make a note of it)

File: src/coinbase/wallet.ts [hunk 7]
```diff
@@ -130,31 +136,67 @@ export class Wallet {
...
+        networkId: Coinbase.networks.BaseSepolia,
```
Comment (0xRAG) [inline src/coinbase/wallet.ts:171]: Should `networkId` be parameterized?

File: src/coinbase/wallet.ts [hunk 8]
```diff
@@ -170,6 +212,32 @@ export class Wallet {
...
+      intervalSeconds,
+    });
+  }
+
+  /**
+   * Creates a new Wallet with the given seed.
+   *
+   * @param options - The options to create the Wallet.
+   * @param options.seed - The seed to use for the Wallet. If undefined, a random seed will be generated.
+   * @param options.networkId - the ID of the blockchain network. Defaults to 'base-sepolia'.
```
Comment (0xRAG) [inline src/coinbase/wallet.ts:228]: Nit: `the` is lowercased here, but uppercased in the other strings

**Review by 0xRAG — APPROVED** [review]
Left some comments & nits, but nothing blocking. 🚢

File: CHANGELOG.md [hunk 9]
```diff
@@ -2,8 +2,12 @@
 
 ## Unreleased
 - Add support for fetching address reputation
-  - Add `reputation` method to `Address` to fetch the reputation of the address. 
+  - Add `reputation` method to `Address` to fetch the reputation of the address
 - Add `networkId` to `WalletData` so that it is saved with the seed data and surfaced via the export function
+- Add ability to import external wallets into CDP via a BIP-39 mnemonic phrase, as a 1-of-1 wallet
+- Add ability to import WalletData files exported by the Python CDP SDK
+- Deprecate `Wallet.loadSeed()` method in favor of `Wallet.loadSeedFromFile()`
+- Deprecate `Wallet.saveSeed()` method in favor of `Wallet.saveSeedToFile()`
```
Comment (alex-stone) [inline CHANGELOG.md:10]: We'll move these around as we prep the release, but we'll have a separate heading for "Address" and "Deprecated"

File: src/coinbase/wallet.ts [hunk 10]
```diff
@@ -130,31 +136,67 @@ export class Wallet {
   }
 
   /**
-   * Imports a Wallet for the given Wallet data.
+   * Loads an existing CDP Wallet using a wallet data object or mnemonic seed phrase.
```
Comment (alex-stone) [inline src/coinbase/wallet.ts:139]: Can we document that it creates a new wallet for mnemonic seed phrases?

Or how can we best clarify that this leaves the existing behavior + adds new behavior, and lays the groundwork for the breaking change in the next release?

Not a blocker per se, but we should aim to make this clear in our public docs + here when we make the cutover

**Review by alex-stone — APPROVED** [review]
Looks good, I think having a big rev on the docs around wallet import v.s. wallet load, etc... is going to be important in the next 2 releases

**Pull request merged**